crates/wasmtime/src/runtime/gc/enabled/structref.rs

```diff
@@ -0,0 +1,757 @@
+        self._ty(store.as_context().0)
+    }
+
+    pub(crate) fn _ty(&self, store: &StoreOpaque) -> Result<StructType> {
+        assert!(self.comes_from_same_store(store));
+        let index = self.type_index(store)?;
+        Ok(StructType::from_shared_type_index(store.engine(), index))
+    }
+
+    /// Does this `structref` match the given type?
+    ///
+    /// That is, is this struct's type a subtype of the given type?
+    ///
+    /// # Errors
+    ///
+    /// Return an error if this reference has been unrooted.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this reference is associated with a different store or if the
+    /// type is not associated with the store's engine.
+    pub fn matches_ty(&self, store: impl AsContext, ty: &StructType) -> Result<bool> {
+        self._matches_ty(store.as_context().0, ty)
+    }
+
+    pub(crate) fn _matches_ty(&self, store: &StoreOpaque, ty: &StructType) -> Result<bool> {
+        assert!(self.comes_from_same_store(store));
+        Ok(self._ty(store)?.matches(ty))
+    }
+
+    pub(crate) fn ensure_matches_ty(&self, store: &StoreOpaque, ty: &StructType) -> Result<()> {
+        if !self.comes_from_same_store(store) {
+            bail!("function used with wrong store");
+        }
+        if self._matches_ty(store, ty)? {
+            Ok(())
+        } else {
+            let actual_ty = self._ty(store)?;
+            bail!("type mismatch: expected `(ref {ty})`, found `(ref {actual_ty})`")
+        }
+    }
+
+    /// Get the values of this struct's fields.
+    ///
+    /// Note that `i8` and `i16` field values are zero-extended into
+    /// `Val::I32(_)`s.
+    ///
+    /// # Errors
+    ///
+    /// Return an error if this reference has been unrooted.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this reference is associated with a different store.
+    pub fn fields<'a, T: 'a>(
+        &self,
+        store: impl Into<StoreContextMut<'a, T>>,
+    ) -> Result<impl ExactSizeIterator<Item = Val> + 'a> {
+        self._fields(store.into().0)
+    }
+
+    pub(crate) fn _fields<'a>(
+        &self,
+        store: &'a mut StoreOpaque,
+    ) -> Result<impl ExactSizeIterator<Item = Val> + 'a> {
+        assert!(self.comes_from_same_store(store));
+
+        let store = AutoAssertNoGc::new(store);
+        let gc_ref = self.inner.try_gc_ref(&store)?.unchecked_copy();
+
+        let header = store.gc_store()?.header(&gc_ref);
+        debug_assert!(header.kind().matches(VMGcKind::StructRef));
+
+        let index = header.ty().expect("structrefs should have concrete types");
+        let ty = StructType::from_shared_type_index(store.engine(), index);
+
+        let layout = store
+            .engine()
+            .signatures()
+            .layout(index)
+            .expect("struct types should have GC layouts");
+        let layout = match layout {
+            crate::vm::GcLayout::Array(_) => unreachable!(),
+            crate::vm::GcLayout::Struct(s) => s,
+        };
```

Comment:
Much of this prefix is shared between here and `set`, would it be possible to refactor that to a shared helper?

Also it feels a little odd where reading fields is always done with the iterator here and `field` is defined in terms of the iterator while `set_field` is defined a field-at-a-time. Could the implementation here delegate to `field` instead of the other way around?